components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt

```diff
@@ -18,89 +22,250 @@ import android.os.ParcelFileDescriptor
+                                currentDownloadJobState.foregroundServiceId.toString()
+                        )
+                    }
+                    ACTION_TRY_AGAIN -> {
+                        currentDownloadJobState.job = startDownloadJob(currentDownloadJobState.state, false)
+                        currentDownloadJobState.status = DownloadJobStatus.ACTIVE
+                        displayOngoingDownloadNotification(currentDownloadJobState.state)
+                    }
+                    ACTION_OPEN -> {
+                        // Create a new file with the location of the saved file to extract the correct path
+                        // `file` has the wrong path, so we must construct it based on the `fileName` and `dir.path`s
+                        val fileLocation = File(currentDownloadJobState.state.filePath ?: "")
+                        val filePath = FileProvider.getUriForFile(
+                                context,
+                                context.packageName + FILE_PROVIDER_EXTENSION,
+                                fileLocation
+                        )
+
+                        val newIntent = Intent(ACTION_VIEW).apply {
+                            setDataAndType(filePath, currentDownloadJobState.state.contentType ?: "*/*")
+                            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_GRANT_READ_URI_PERMISSION
+                        }
+
+                        startActivity(newIntent)
+                    }
+                }
+            }
+        }
     }
 
     override fun onBind(intent: Intent?): IBinder? = null
 
-    override suspend fun onStartCommand(intent: Intent?, flags: Int) {
-        val download = intent?.getDownloadExtra() ?: return
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        val download = intent?.getDownloadExtra() ?: return START_REDELIVER_INTENT
+        registerForUpdates()
 
-        val notification = try {
-            performDownload(download)
-            DownloadNotification.createDownloadCompletedNotification(context, download.fileName)
-        } catch (e: IOException) {
-            DownloadNotification.createDownloadFailedNotification(context, download.fileName)
+        val foregroundServiceId = Random.nextInt()
+
+        // Create a new job and add it, with its downloadState to the map
+        listOfDownloadJobs[download.id] = DownloadJobState(
+            job = startDownloadJob(download, false),
+            state = download,
+            foregroundServiceId = foregroundServiceId,
+            status = DownloadJobStatus.ACTIVE
+        )
+
+        return super.onStartCommand(intent, flags, startId)
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        listOfDownloadJobs.values.forEach {
+            it.job?.cancel()
         }
+    }
+
+    private fun startDownloadJob(download: DownloadState, isResuming: Boolean): Job {
+        return CoroutineScope(IO).launch {
```

Comment:
If you have a scope for the service, you could just `cancel` that rather than having to loop through jobs in `onDestroy`

Reply:
Where should I cancel, then?

Reply:
```kt
class Service {
  private val scope = CoroutineScope(IO)

  fun onStartCommand() {
    scope.launch { startDownloadJob() }
  }

  fun onDestroy() {
    scope.cancel()
  }
}